src/kmedoids_pywrapper.cpp

```diff
@@ -38,116 +39,34 @@ int sum_thread_ids() {
   m.def("set_num_threads", &omp_set_num_threads, "Set the maximum number of threads");
   m.def("sum_thread_ids", &sum_thread_ids, "Adds the ids of threads; used only for debugging");
-  py::class_<KMedoidsWrapper>(m, "KMedoids")
-      .def(py::init<int, std::string, int, int, int, int, std::string>(),
-        py::arg("n_medoids") = NULL,
-        py::arg("algorithm") = "BanditPAM",
-        py::arg("verbosity") = 0,
-        py::arg("maxIter") = 1000,
-        py::arg("buildConfidence") = 1000,
-        py::arg("swapConfidence") = 10000,
-        py::arg("logFilename") = "KMedoidsLogfile"
-      )
-      .def_property("n_medoids", &KMedoidsWrapper::getNMedoids, &KMedoidsWrapper::setNMedoids)
-      .def_property("algorithm", &KMedoidsWrapper::getAlgorithm, &KMedoidsWrapper::setAlgorithm)
-      .def_property("verbosity", &KMedoidsWrapper::getVerbosity, &KMedoidsWrapper::setVerbosity)
-      .def_property("maxIter", &KMedoidsWrapper::getMaxIter, &KMedoidsWrapper::setMaxIter)
-      .def_property("buildConfidence", &KMedoidsWrapper::getbuildConfidence, &KMedoidsWrapper::setbuildConfidence)
-      .def_property("swapConfidence", &KMedoidsWrapper::getswapConfidence, &KMedoidsWrapper::setswapConfidence)
-      .def_property("logFilename", &KMedoidsWrapper::getLogfileName, &KMedoidsWrapper::setLogFilename)
-      .def_property_readonly("medoids", &KMedoidsWrapper::getMedoidsFinalPython)
-      .def_property_readonly("build_medoids", &KMedoidsWrapper::getMedoidsBuildPython)
-      .def_property_readonly("labels", &KMedoidsWrapper::getLabelsPython)
-      .def_property_readonly("steps", &KMedoidsWrapper::getStepsPython)
-      .def("fit", &KMedoidsWrapper::fitPython);
+  py::class_<KMedoidsWrapper> cls(m, "KMedoids"); 
+  cls.def(py::init<int, std::string, int, int, int, int, std::string>(),
+          py::arg("n_medoids") = NULL,
+          py::arg("algorithm") = "BanditPAM",
+          py::arg("verbosity") = 0,
+          py::arg("maxIter") = 1000,
+          py::arg("buildConfidence") = 1000,
+          py::arg("swapConfidence") = 10000,
+          py::arg("logFilename") = "KMedoidsLogfile"
+  );
+  cls.def_property("n_medoids", &KMedoidsWrapper::getNMedoids, &KMedoidsWrapper::setNMedoids);
+  cls.def_property("algorithm", &KMedoidsWrapper::getAlgorithm, &KMedoidsWrapper::setAlgorithm);
+  cls.def_property("verbosity", &KMedoidsWrapper::getVerbosity, &KMedoidsWrapper::setVerbosity);
+  cls.def_property("maxIter", &KMedoidsWrapper::getMaxIter, &KMedoidsWrapper::setMaxIter);
+  cls.def_property("buildConfidence", &KMedoidsWrapper::getbuildConfidence, &KMedoidsWrapper::setbuildConfidence);
+  cls.def_property("swapConfidence", &KMedoidsWrapper::getswapConfidence, &KMedoidsWrapper::setswapConfidence);
+  cls.def_property("logFilename", &KMedoidsWrapper::getLogfileName, &KMedoidsWrapper::setLogFilename);
+  medoids_python(cls);
+  build_medoids_python(cls);
+  labels_python(cls);
+  steps_python(cls);
+  fit_python(cls);
```

Comment:
Can you educate me about the syntax of these lines? Why not include the `cls.def_property_readonly`? And how are these linked to their C++ fns? I.e. `&KMedoidsWrapper::getLabelsPython` for `labels`?

Reply:
This is the example I follow: https://pybind11.readthedocs.io/en/stable/faq.html?highlight=reduce%20build%20time#how-can-i-reduce-the-build-time

Take `labels` as an example:

In `labels_python.cpp`, I have defined two functions `&KMedoidsWrapper::getLabelsPython()` and `labels_python()`. Essentially, `labels_python()` applies the `cls.def_property_readonly` to `&KMedoidsWrapper::getLabelsPython()`. That's why we only need to call this function in the binding! Note that all the new functions are defined in the header `kmedoids_pywrapper.hpp`.